src/referenced-fields.ts

```diff
@@ -294,3 +313,271 @@ export const getModifiedFields: EngineInstance['getModifiedFields'] = (
+					);
+				}
+				if (SELECT_QUERY_TYPES.includes(arg[0])) {
+					findBindCandidates(arg, candidates);
+				}
+			}
+			return null;
+		case 'From':
+		case 'CrossJoin':
+			const fromNode = statement as FromNode;
+			return findTableInTableDefinition(fromNode[1], candidates);
+		case 'Join':
+		case 'LeftJoin':
+		case 'RightJoin':
+		case 'FullJoin':
+			const joinNode = statement as InnerJoinNode;
+			if (joinNode[2]) {
+				findTablesInConstraint(joinNode[2][1], candidates);
+			}
+			return findTableInTableDefinition(joinNode[1], candidates);
+		case 'Where':
+			const whereNode = statement as WhereNode;
+			findTablesInConstraint(whereNode[1], candidates);
+			return null;
+		default:
+			return null;
+	}
+};
+
+const findBindCandidates = (
+	statement: AbstractSqlQuery,
+	candidates: BindCandidate[],
+) => {
+	switch (statement[0]) {
+		case 'SelectQuery':
+			const selectQueryNode = statement as SelectQueryNode;
+			const scope = statement;
+			for (const arg of selectQueryNode.slice(1)) {
+				if (!isAbstractSqlQuery(arg)) {
+					throw new Error(
+						'cannot introspect into the string form of AbstractSqlQuery',
+					);
+				}
+				const table = findTableInStatement(arg, candidates);
+				if (table) {
+					candidates.push({ ...table, scope });
+				}
+			}
+			break;
+		default:
+			throw new Error(
+				`findBindCandidates only supports SELECT queries, got ${statement[0]}`,
+			);
+	}
+};
+
+// TODO: lots
+// - Narrowing rows simply by narrowing the IDs to those that were modified on
+//   some table is not safe most of the time. As far as we know, it is safe to
+//   do this with any table selected from one and quantified with a forall or
+//   not exists. Finding these cases is not trivial as, for example,
+//   "not not exists" is in the end in existential which is not safe with this
+//   method. Because of this we restrict ourselves to narrowing only the root
+//   table right now.
+// - Removing multiple candidates selecting from the same database table to
+//   avoid visibility issues is too conservative. The correct criteria is to
+//   just remove any that are present in at least 2 disjoint subqueries.
+//   The problem is that in those cases there is no single place in the query
+//   that has visibility inside both disjoint subqueries and that is a
+//   requirement for adding the corrent binds for narrowing.
+// - We can miss subqueries as we don't recurse into everything.
+// - Ban on aggregates is not necessary but solving it requires additional
+//   complexity.
+// - We only care about and recognize `COUNT` aggregates and `GROUP BY`
+//   statements right now.
+// - We assume the ID column is named "id".
+// - There are other techniques that cover cases that this simple narrowing
+//   does not. Currently that are two with their own kinks to work out:
+//   ID propagation and COUNT splitting.
+export const addAffectedIdsBinds = (
+	abstractSql: AbstractSqlQuery,
```

Comment:
Please add some comment that the input data is manipulated in situ / in place.

Reply:
I would think that the name itself `addAffectedIdsBinds` would be enough to show that, specially because this function doesn't return anything. Maybe `decorate` is better than `add` to convey this?